keylime/config.py

```diff
@@ -73,45 +74,257 @@ def environ_bool(env_name, default):
-                file = resources.files(__package__).joinpath("keylime.conf")
-                config.read_string(file.read_text("utf-8"))
+                file = resources.files(__package__).joinpath(f"config/{component}.conf")
+                _config[component].read_string(file.read_text("utf-8"))
         else:
-            # Validate that at least one config file is present
-            config_files = config.read(CONFIG_FILES)
-            # TODO - use the logger
-            print(f"Reading configuration from {config_files}")
-        get_config.config = config
-    return get_config.config
+            for c in CONFIG_FILES[component]:
+                # Search for configuration file in order of priority given by
+                # CONFIG_FILES. The first base configuration file found is used,
+                # the others are ignored
+
+                # Validate that at least one config file is present
+                config_file = _config[component].read(c)
+                # TODO - use the logger
+                if config_file:
+                    print(f"Reading configuration from {config_file}")
+
+                    if not CONFIG_SNIPPETS_DIRS or not isinstance(CONFIG_SNIPPETS_DIRS, dict):
+                        raise Exception("Invalid CONFIG_FILES")
+
+                    if not component in CONFIG_SNIPPETS_DIRS:
+                        raise Exception(f"Invalid component {component}")
+
+                    for d in (x for x in CONFIG_SNIPPETS_DIRS[component] if os.path.exists(x)):
+                        snippets = sorted(filter(os.path.isfile, (os.path.join(d, f) for f in os.listdir(d) if f)))
+                        applied_snippets = _config[component].read(snippets)
+                        if applied_snippets:
+                            # TODO - use the logger
+                            print(f"Applied configuration snippets from {d}")
+                    break
+
+    return _config[component]
+
+
+def getlist(component, option, section=None):
+    if not section:
+        section = component
+
+    read = get_config(component).get(section, option)
+
+    if read:
+        try:
+            l = ast.literal_eval(read)
+            if isinstance(l, list):
+                stripped = []
+                for s in l:
+                    if isinstance(s, str):
+                        s = s.strip()
+                    stripped.append(s)
+                return stripped
+            raise Exception(
+                f"Config option '{option}' in section '{section}' " f"'of component {component} should be a list"
+            )
+        except Exception as e:
+            raise Exception(
+                f"Failed to get list from config for component '{component}', section '{section}', option '{option}'"
+            ) from e
+
+    raise Exception(f"Could not find option '{option}' in section '{section}' of component '{component}'")
+
+
+def getdict(component, option, section=None):
```

Comment:
This is not used? Do we require it?

Reply:
The `tpm_policy` was a dictionary, indeed. But it was dropped from the configuration file , following the enhancement proposal:

https://github.com/keylime/enhancements/blob/master/72_config_and_simplify_tls.md#removal-of-unused-default-values-from-the-configuration-file

The default  `tpm_policy` used by the tenant is the empty policy (`'{}'`)

I implemented this as the enhancement proposal asked for it:

https://github.com/keylime/enhancements/blob/master/72_config_and_simplify_tls.md#unification-of-data-types

In the end it is not used and I'll drop it.